modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GcsRepositoryStatsCollector.java

```diff
@@ -0,0 +1,282 @@
+        setThreadLocal(stats);
+        var t = timer.absoluteTimeInMillis();
+        try {
+            return blobFn.get();
+        } finally {
+            stats.totalDuration += timer.absoluteTimeInMillis() - t;
+            clearThreadLocal();
+        }
+    }
+
+    /**
+     * Final step in continual collection
+     */
+    public void finishRunnable(OperationStats stats, IORunnable runnable) throws IOException {
+        setThreadLocal(stats);
+        var t = timer.absoluteTimeInMillis();
+        try {
+            runnable.run();
+        } finally {
+            stats.totalDuration += timer.absoluteTimeInMillis() - t;
+            collect(stats);
+            clearThreadLocal();
+        }
+    }
+
+    /**
+     * Continue collecting metrics with given OperationStats. Useful for readers and writers.
+     */
+    public void collectIORunnable(OperationPurpose purpose, StorageOperation operation, IORunnable runnable) throws IOException {
+        var stats = initAndGetThreadLocal(purpose, operation);
+        var t = timer.absoluteTimeInMillis();
+        try {
+            runnable.run();
+        } finally {
+            stats.totalDuration += timer.absoluteTimeInMillis() - t;
+            clearThreadLocal();
+        }
+    }
+
+    public void collectRunnable(OperationPurpose purpose, StorageOperation operation, Runnable runnable) {
+        var t = timer.absoluteTimeInMillis();
+        var stats = initAndGetThreadLocal(purpose, operation);
+        try {
+            runnable.run();
+        } finally {
+            stats.totalDuration += timer.absoluteTimeInMillis() - t;
+            collect(stats);
+            clearThreadLocal();
+        }
+    }
+
+    /**
+     * Executes GCS Storage operation in a wrapper that stores metrics in ThreadLocal
+     */
+    public <T> T collectIOSupplier(OperationPurpose purpose, StorageOperation operation, IOSupplier<T> blobFn) throws IOException {
+        var t = timer.absoluteTimeInMillis();
+        var stats = initAndGetThreadLocal(purpose, operation);
+        try {
+            return blobFn.get();
+        } finally {
+            stats.totalDuration += timer.absoluteTimeInMillis() - t;
+            collect(stats);
+            clearThreadLocal();
+        }
+    }
+
+    /**
+     * Executes GCS Storage operation in a wrapper that stores metrics in ThreadLocal
+     */
+    public <T> T collectSupplier(OperationPurpose purpose, StorageOperation operation, Supplier<T> blobFn) {
+        var t = timer.absoluteTimeInMillis();
+        var stats = initAndGetThreadLocal(purpose, operation);
+        try {
+            return blobFn.get();
+        } finally {
+            stats.totalDuration += timer.absoluteTimeInMillis() - t;
+            collect(stats);
+            clearThreadLocal();
+        }
+    }
+
+    private void collect(OperationStats stats) {
+        if (stats.reqAtt == 0) {
```

Comment:
I kinda think we shouldn't use abbreviations here, `requestsAttempted` reads much nicer than `Att` especially because `att` can mean so many things

Reply:
And we have attributes floating around in the same block of code. It's a bit confusing.